embassy-usb-dfu/src/dfu.rs

```diff
@@ -60,6 +60,21 @@ impl<'d, DFU: NorFlash, STATE: NorFlash, RST: Reset, const BLOCK_SIZE: usize> Ha
             }
             Ok(Request::Dnload) if self.attrs.contains(DfuAttributes::CAN_DOWNLOAD) => {
                 if req.value == 0 {
+                    match self.updater.prepare_update() {
```

Comment:
I think rather than introducing write_firmware_without_erase API, you should use the flash value returned from prepare_update to do the writing. Or is there something preventing that?

Reply:
That was the first thing I tried, but I struggled with lifetimes. It was especially tricky when I had both a `BootLoader` and a `BlockingFirmwareUpdater` instances in different scopes of the bootloader. For example, if the bootloader starts and it detects that a specific button is pressed, it sets `State::DfuDetach` and begins a USB DFU update.